compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -1788,6 +1788,49 @@ impl<'a> Parser<'a> {
             self.sess.expr_parentheses_needed(&mut err, *sp);
         }
         err.span_label(span, "expected expression");
+        if self.prev_token.kind == TokenKind::BinOp(token::Plus)
+            && self.token.kind == TokenKind::BinOp(token::Plus)
+            && self.look_ahead(1, |t| !t.is_lit())
+        {
+            let span = self.prev_token.span.to(self.token.span);
+            err.note("Rust has no dedicated increment operator");
+            err.span_suggestion_verbose(
+                span,
+                "try using `+= 1` instead",
+                " += 1".into(),
+                Applicability::Unspecified,
+            );
+        } else if self.token.kind == TokenKind::BinOp(token::Plus)
+            && self.look_ahead(1, |t| t.kind == TokenKind::BinOp(token::Plus))
+            && self.look_ahead(2, |t| !t.is_lit())
+        {
+            let target_span = self.look_ahead(2, |t| t.span);
+            let left_brace_span = target_span.shrink_to_lo();
+            let pre_span = self.token.span.to(self.look_ahead(1, |t| t.span));
+            let post_span = target_span.shrink_to_hi();
+
+            err.note("Rust has no dedicated increment operator");
+
+            if self.prev_token.kind == TokenKind::Semi {
```

Comment:
This seems to be a good heuristic so that we don't suggest `{ i += 1; i};`. However, I wonder if there are places where this heuristic yields a false positive, and we emit code that is different from what the user meant. It seems unlikely, but should we change the applicability for this version of the suggestion to `MaybeIncorrect` just in case?

Reply:
Looking at all these changes, I feel like a better place to put this is in `parse_lit_expr`. That way we can

* give a specific error instead of the generic "expected expression":

```
error: Rust has no postfix increment operator
  --> $DIR/increment.rs:3:7
   |
LL |     i++;
   |      ^^ not a valid unary operator
   |
help: use in-place increment instead
   |
LL |     i += 1;
   |       ~~~~
```

* we can recover the parser and return that as an expression (which would also give you mutability and type errors if appropriate)
* gives us a good place where we can look for the prefix version: you add a `None if matches!(self.token.kind, Plus | Minus)` (and a lookahead 1 and 2 to check for another `Plus | Minus` and for a numeric literal or ident) case in the match in that function
* you can still check for whether the next thing will be a `Semi` to not give the `{}` surrounded version

`parse_lit_expr` is already the fall-through of `parse_bottom_expr`, which is effectively where expressions are parsed.

Reply:
> you can still check for whether the next thing will be a `Semi` to not give the `{}` surrounded version

Actually I need to check for whether the *previous* thing was a `Semi`, because otherwise on code like this:

```rust
let y = x++;

we will suggest this:

```rust
let y = x += 1;
```

How should I get around this? Will the `prev_token` in `parse_lit_expr` work?

Reply:
It looks like by the time `parse_lit_expr` is invoked for the postfix version, we've lost the context of the previous `;` because the `prev_token` is the variable being incremented. What should I do?

Reply:
Actually, it looks like `prev_token` is the first `+`, not even the variable. This seems like a problem...

Reply:
I moved the code to `parse_assoc_expr_with()` (and extracted a bunch of helper functions), which has more information.